### PR TITLE
[RFC] Events keywords

### PR DESCRIPTION
**Pull request by astronouth7303**
In discussing the future of command transformation, I realized that history backends might want to log the pre- and post-transformation commands. But changing the signature of `on_precommand` will break  existing users.

So, moving forward, I propose that events are keyword-argument only, and that all handlers have a `**kw` against future expansion. (There is precedent for this. Django--and the PyDispatch it was based on--does the same thing.)

* [x] Implement in core usage
* [x] Update documentation
* [x] News

### CODE REVIEW

**Comment by laerus** [comment]
+1

File: xonsh/base_shell.py [hunk 1]
```diff
@@ -375,7 +375,7 @@ def _fix_cwd(self):
             builtins.__xonsh_env__['PWD'] = cwd      # track it now
             if old is not None:
                 builtins.__xonsh_env__['OLDPWD'] = old  # and update $OLDPWD like dirstack.
-            events.on_chdir.fire(old, cwd)              # fire event after cwd actually changed.
+            events.on_chdir.fire(olddir=old, newdir=cwd)              # fire event after cwd actually changed.
```
Comment (laerus) [inline xonsh/base_shell.py]: maybe just `old` `new` ? (naming itches :p)

File: xonsh/events.py [hunk 2]
```diff
@@ -49,37 +63,41 @@ def __call__(self, handler):
         rtn : callable
             The handler
         """
-        #  Using Pythons "private" munging to minimize hypothetical collisions
+        #  Using Python's "private" munging to minimize hypothetical collisions
         handler.__validator = None
+        if debug_level():
+            if not has_kwargs(handler):
```
Comment (laerus) [inline xonsh/events.py]: shouldnt this check also run when `XONSH_DEBUG=0`?

File: xonsh/events.py [hunk 3]
```diff
@@ -8,11 +8,25 @@
...
 
 
+def has_kwargs(func):
+    return any(p.kind == p.VAR_KEYWORD for p in inspect.signature(func).parameters.values())
+
+
+def debug_level():
+    if hasattr(builtins, '__xonsh_env__'):
+        return builtins.__xonsh_env__.get('XONSH_DEBUG')
+    # FIXME: Under py.test, return 1(?)
```
Comment (laerus) [inline xonsh/events.py]: i dont think this module should worry about pytest

**Comment by scopatz** [comment]
This looks great to me! Thanks a million @astronouth7303!

**Comment by scopatz** [comment]
Thanks for reviewing @laerus!

File: xonsh/base_shell.py (hunk 1 above)
Comment (astronouth7303) [inline xonsh/base_shell.py]: `new` always feels like a keyword to me, even if it isn't in Python.

File: xonsh/events.py (hunk 3 above)
Comment (astronouth7303) [inline xonsh/events.py]: Shouldn't, but if you've got a better way to deal with it, I'm ears.

File: xonsh/events.py (hunk 2 above)
Comment (astronouth7303) [inline xonsh/events.py]: It's not required to run, and does technically slow things down. And it's not like end-users could do something about it.

**Pull request merged**